File: python/fwd_client/idempotency.py
"""Generic idempotency-key helper for fwd callers.

fwd deduplicates signing requests by ``Idempotency-Key`` header (≤128 chars).
The helpers here produce stable, bounded keys from arbitrary string parts —
they carry no clif-specific knowledge (no network table, no claim logic).

Caller-specific composition (e.g. ``f"clif:{network}:{claim_type}:…"``) lives
in the consumer, not here.
"""

from __future__ import annotations

import hashlib

_MAX_KEY_LEN: int = 128


def make_idempotency_key(*parts: str, retry: str | None = None) -> str:
    """Deterministic sha256-based idempotency key from arbitrary string parts.

    The key is stable across processes: the same *parts* (and the same
    optional *retry* discriminator) always produce the same key, so a network
    retry or crash-rerun of the **same logical attempt** dedups at fwd.

    ``retry`` is the operator-controlled discriminator for a **deliberate**
    logical re-attempt after an on-chain failure (fwd replay is status-blind
    by design).  ``retry=None`` ⇒ the key is byte-identical to a key produced
    without the retry argument.  A new ``retry`` value ⇒ a fresh key.  Callers
    should never auto-generate ``retry`` — it is set explicitly by the operator.

    The returned key is ≤128 chars (fwd's limit).
    """
    raw = ":".join(parts)
    if retry is not None:
        raw += f":retry={retry}"
    digest = hashlib.sha256(raw.encode()).hexdigest()
    # Prefix with sanitised parts (first 80 chars) so the key is human-readable
    # in logs; the digest suffix guarantees uniqueness regardless of prefix length.
    prefix = raw[:80].replace(" ", "_")
    key = f"{prefix}-{digest[:16]}"
    # Truncate to the hard limit (prefix may be shorter than 80 chars, so in
    # practice this only fires on very long composite parts).
    return key[:_MAX_KEY_LEN]

File: go/idempotency.go
package fwdclient

import (
	"crypto/sha256"
	"encoding/hex"
	"strings"
)

const maxKeyLen = 128

// MakeIdempotencyKey is a byte-identical port of the Python fwd-client
// make_idempotency_key: a deterministic, bounded (<=128 char) key derived from
// arbitrary string parts, so a network retry or crash-rerun of the SAME logical
// attempt dedups at fwd. It carries no consumer-specific knowledge — composition
// (e.g. "clif:<network>:<type>:<epoch>") lives in the consumer.
//
// retry is the operator-controlled discriminator for a DELIBERATE re-attempt
// after an on-chain failure. An empty retry ("") means no discriminator and is
// byte-identical to the Python retry=None case; a non-empty retry appends
// ":retry=<retry>" (Python retry=<value>). For identical (parts, retry) inputs,
// Go and Python produce identical keys. (The degenerate Python retry="" case —
// append ":retry=" — is not representable here; use a non-empty tag.)
func MakeIdempotencyKey(parts []string, retry string) string {
	raw := strings.Join(parts, ":")
	if retry != "" {
		raw += ":retry=" + retry
	}
	sum := sha256.Sum256([]byte(raw))
	digest := hex.EncodeToString(sum[:]) // 64 lowercase hex chars
	// Human-readable prefix (first 80 chars of raw, spaces -> underscores) plus
	// the first 16 hex chars of the digest for uniqueness. Slicing is rune-based
	// to match Python str slicing (identical to byte-slicing for ASCII inputs).
	prefix := strings.ReplaceAll(truncRunes(raw, 80), " ", "_")
	key := prefix + "-" + digest[:16]
	return truncRunes(key, maxKeyLen)
}

// truncRunes returns the first n runes of s (Python str[:n] semantics).
func truncRunes(s string, n int) string {
	if len(s) <= n { // fast path: byte len <= n implies rune len <= n
		return s
	}
	r := []rune(s)
	if len(r) <= n {
		return s
	}
	return string(r[:n])
}
